File: wallet/android/app/src/main/kotlin/ai/banya/linkcpp/wallet/NodeSettings.kt
package ai.banya.linkcpp.wallet

import android.os.Build
import android.provider.Settings
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import kotlinx.coroutines.launch
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import ai.banya.linkcpp.core.StakingService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext

private val AMBER = Color(0xFFE0952B)
private val GREEN = Color(0xFF33C06B)

@Composable
fun NodeSettingsScreen(vm: WalletViewModel, nav: NavController) {
    val b = LocalBrand.current
    val s = LocalStrings.current
    val ctx = LocalContext.current
    val profile = nodeProfile(vm.nodeBackend, vm.nodeMode, s)
    var stats by remember { mutableStateOf(DeviceStats()) }

    LaunchedEffect(vm.nodeLive, vm.stakingUrl) {
        if (!vm.nodeLive) return@LaunchedEffect
        val owner = vm.address ?: return@LaunchedEffect
        val svc = StakingService(vm.stakingUrl)
        val accel = if (vm.nodeBackend == "cpu") "cpu" else "gpu"
        val nodeId = "android-${(Settings.Secure.getString(ctx.contentResolver, Settings.Secure.ANDROID_ID) ?: "dev").take(8)}"
        NodeTelemetry.read(ctx) // prime CPU delta
        runCatching {
            withContext(Dispatchers.IO) {
                svc.registerNode(nodeId, owner, os = "android", deviceKind = "phone",
                    accelerator = accel, label = "${Build.MODEL} · ${vm.nodeMode}",
                    perfScore = profile.tokPerSec.toDouble(), backend = vm.nodeBackend, mode = vm.nodeMode)
            }
        }
        while (vm.nodeLive) {
            stats = NodeTelemetry.read(ctx)
            runCatching { withContext(Dispatchers.IO) { svc.heartbeat(nodeId) } }
            delay(1500)
        }
    }

    BrandBackground {
        Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(20.dp)) {
            ScreenHeader(s.t("ns.title"), nav)

            // backend
            Column(Modifier.brandCard()) {
                Text(s.t("ns.backend"), color = b.textPrimary, fontSize = 16.sp, fontWeight = FontWeight.Bold)
                Spacer(Modifier.height(4.dp))
                Text(s.t("ns.backendDesc"), color = b.textSecondary, fontSize = 12.sp)
                Spacer(Modifier.height(10.dp))
                Row {
                    ChoiceChip("GPU · OpenCL", vm.nodeBackend == "opencl") { vm.updateNodeBackend("opencl") }
                    Spacer(Modifier.width(8.dp))
                    ChoiceChip("GPU · Vulkan", vm.nodeBackend == "vulkan") { vm.updateNodeBackend("vulkan") }
                    Spacer(Modifier.width(8.dp))
                    ChoiceChip("CPU", vm.nodeBackend == "cpu") { vm.updateNodeBackend("cpu") }
                }
            }
            Spacer(Modifier.height(12.dp))

            // mode
            Column(Modifier.brandCard()) {
                Text(s.t("ns.mode"), color = b.textPrimary, fontSize = 16.sp, fontWeight = FontWeight.Bold)
                Spacer(Modifier.height(10.dp))
                ModeRow(s.t("ns.localShard"), s.t("ns.localShardDesc"),
                    vm.nodeMode == "local_shard") { vm.updateNodeMode("local_shard") }
                Spacer(Modifier.height(8.dp))
                ModeRow(s.t("ns.rpcWorker"), s.t("ns.rpcWorkerDesc"),
                    vm.nodeMode == "rpc_worker") { vm.updateNodeMode("rpc_worker") }
                if (vm.nodeLive && vm.nodeMode == "rpc_worker") {
                    Spacer(Modifier.height(8.dp))
                    Text(
                        if (vm.nodeAgentRunning) "RPC worker listening · agent :${vm.nodeAgentPort}"
                        else vm.nodeAgentEvent.ifEmpty { "starting…" },
                        color = if (vm.nodeAgentRunning) GREEN else b.textSecondary,
                        fontSize = 11.sp, fontFamily = androidx.compose.ui.text.font.FontFamily.Monospace,
                    )
                }
            }
            Spacer(Modifier.height(12.dp))

            // infographic
            Column(Modifier.brandCard()) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(s.t("ns.resImpact"), color = b.textPrimary, fontSize = 16.sp, fontWeight = FontWeight.Bold)
                    Spacer(Modifier.weight(1f))
                    Column(horizontalAlignment = Alignment.End) {
                        Text("${profile.tokPerSec}", style = androidx.compose.ui.text.TextStyle(brush = brandGradient(), fontSize = 24.sp, fontWeight = FontWeight.ExtraBold))
                        Text("tok/s (0.5B Q8)", color = b.textSecondary, fontSize = 10.sp)
                    }
                }
                Text(profile.computeUnit, color = b.blue, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
                Spacer(Modifier.height(10.dp))
                MetricBar(s.t("ns.memImpact"), profile.memImpact, "${(profile.memImpact * 100).toInt()}%", b.blue)
                MetricBar(s.t("ns.thermal"), profile.thermal, thermalWord(profile.thermal, s), AMBER)
                MetricBar(s.t("ns.performance"), profile.performance, "${(profile.performance * 100).toInt()}%", GREEN)
                Spacer(Modifier.height(6.dp))
                Text("※ ${profile.note}", color = b.textSecondary, fontSize = 11.sp)
            }
            Spacer(Modifier.height(12.dp))

            // policy + live
            Column(Modifier.brandCard()) {
                ToggleRow(s.t("ns.chargingOnly"), vm.nodeChargingOnly) { vm.updateNodeChargingOnly(it) }
                Spacer(Modifier.height(8.dp))
                ToggleRow(s.t("ns.runLive"), vm.nodeLive) { vm.updateNodeLive(it) }
            }

            Spacer(Modifier.height(12.dp))
            HubConnectCard(vm)

            if (vm.nodeLive) {
                Spacer(Modifier.height(12.dp))
                Column(Modifier.brandCard()) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Box(Modifier.size(9.dp).clip(RoundedCornerShape(50)).background(GREEN))
                        Spacer(Modifier.width(8.dp))
                        Text(s.t("ns.liveGauges"), color = b.textPrimary, fontSize = 16.sp, fontWeight = FontWeight.Bold)
                        Spacer(Modifier.weight(1f))
                        Text(if (stats.charging) s.t("ns.charging") else s.t("ns.onBattery"), color = b.textSecondary, fontSize = 12.sp)
                    }
                    Spacer(Modifier.height(12.dp))
                    val ramFrac = if (stats.ramTotalMb > 0) stats.ramUsedMb.toFloat() / stats.ramTotalMb else 0f
                    Gauge("RAM", ramFrac, "%.1f / %.1f GB".format(stats.ramUsedMb / 1024f, stats.ramTotalMb / 1024f), b.blue, true)
                    Gauge(s.t("ns.cpuLoad"), stats.cpuLoad, "${(stats.cpuLoad * 100).toInt()}%", GREEN, true)
                    val tempFrac = ((stats.batteryTempC - 20f) / 30f).coerceIn(0f, 1f)
                    Gauge(s.t("ns.temp"), tempFrac, "%.1f°C".format(stats.batteryTempC), AMBER, true)
                    Gauge("GPU (${vm.nodeBackend})", profile.performance, s.t("ns.estimated"), b.pink, false)
                    Gauge("NPU", 0f, s.t("ns.unused"), b.textSecondary, false)
                    Spacer(Modifier.height(4.dp))
                    Text(s.t("ns.gaugeNote"), color = b.textSecondary, fontSize = 10.sp)
                }
            }
            Spacer(Modifier.height(12.dp))
            ApiKeyCard(vm)
            Spacer(Modifier.height(24.dp))
        }
    }
}

/**
 * Reissue the credit API key.
 *
 * The key is stored on this device and the gateway keeps only its hash, so a key
 * it no longer recognises cannot be repaired by asking for it back — it has to
 * be reminted. Without this, "invalid API key" has no way out from inside the
 * app. Credit balance is held against the wallet, not the key.
 */
@Composable
private fun ApiKeyCard(vm: WalletViewModel) {
    val b = LocalBrand.current
    val s = Strings(vm.language)
    val scope = rememberCoroutineScope()
    var busy by remember { mutableStateOf(false) }
    var status by remember { mutableStateOf("") }

    Column(Modifier.brandCard()) {
        Text(s.t("ns.apiKey"), color = b.textPrimary, fontSize = 16.sp, fontWeight = FontWeight.Bold)
        Spacer(Modifier.height(4.dp))
        Text(s.t("ns.apiKeyDesc"), color = b.textSecondary, fontSize = 12.sp)
        Spacer(Modifier.height(10.dp))
        Box(
            Modifier.fillMaxWidth()
                .clip(RoundedCornerShape(10.dp))
                .background(if (busy) b.pink.copy(alpha = 0.4f) else b.pink)
                .clickableNoRipple {
                    if (busy) return@clickableNoRipple
                    busy = true; status = s.t("ns.apiKeyWorking")
                    scope.launch {
                        val r = runCatching { vm.reissueCreditApiKey() }
                        status = r.fold({ s.t("ns.apiKeyOk") }, { "${s.t("ns.apiKeyFailed")}: ${it.message}" })
                        busy = false
                    }
                }
                .padding(vertical = 12.dp), contentAlignment = Alignment.Center,
        ) { Text(if (busy) s.t("ns.apiKeyWorking") else s.t("ns.apiKeyReissue"), color = Color.White, fontWeight = FontWeight.SemiBold, fontSize = 13.sp) }

        if (status.isNotEmpty()) {
            Spacer(Modifier.height(8.dp))
            Text(status, color = if (status == s.t("ns.apiKeyOk")) GREEN else b.textSecondary, fontSize = 11.sp)
        }
    }
}

private fun thermalWord(v: Float, s: Strings): String = when {
    v >= 0.75f -> s.t("thermal.high"); v >= 0.5f -> s.t("thermal.medium"); else -> s.t("thermal.low")
}

/**
 * Connect this node to a remote (auth-gated) hub: sign in with the wallet
 * (SIWS) + OTP, then hand the resulting bearer token to the running node agent
 * so it polls that hub's shard-demand market. The LAN hub is discovered
 * automatically; this is for public hubs the node can only reach outbound.
 */
@Composable
private fun HubConnectCard(vm: WalletViewModel) {
    val b = LocalBrand.current
    val ctx = LocalContext.current
    val scope = rememberCoroutineScope()
    var url by remember { mutableStateOf("https://hub.kvasir-ai.net") }
    var busy by remember { mutableStateOf(false) }
    var status by remember { mutableStateOf("") }
    val hubs = remember { mutableStateOf(NodeService.agent?.knownHubList() ?: emptyList()) }

    Column(Modifier.brandCard()) {
        Text("허브 연결", color = b.textPrimary, fontSize = 16.sp, fontWeight = FontWeight.Bold)
        Spacer(Modifier.height(4.dp))
        Text("원격 허브에 지갑 서명으로 노드 토큰을 발급받아 폴링 등록 (OTP 불필요)", color = b.textSecondary, fontSize = 12.sp)
        Spacer(Modifier.height(10.dp))
        androidx.compose.material3.OutlinedTextField(
            value = url, onValueChange = { url = it }, singleLine = true,
            label = { Text("허브 URL", fontSize = 12.sp) },
            modifier = Modifier.fillMaxWidth(),
        )
        Spacer(Modifier.height(10.dp))
        Box(
            Modifier.fillMaxWidth()
                .clip(RoundedCornerShape(10.dp))
                .background(if (busy) b.pink.copy(alpha = 0.4f) else b.pink)
                .clickableNoRipple {
                    if (busy) return@clickableNoRipple
                    val words = vm.revealMnemonic()
                    if (words == null) { status = "지갑 잠금 해제 필요"; return@clickableNoRipple }
                    if (NodeService.agent == null) { status = "노드를 먼저 켜세요 (라이브)"; return@clickableNoRipple }
                    busy = true; status = "지갑 서명 중…"
                    scope.launch {
                        val r = withContext(Dispatchers.IO) { runCatching { HubAuthService(url, words).nodeToken() } }
                        r.onSuccess {
                            NodeService.agent?.registerHub(url, it)
                            hubs.value = NodeService.agent?.knownHubList() ?: emptyList()
                            status = "연결됨 · 노드 토큰 등록"
                        }.onFailure { status = "실패: ${it.message}" }
                        busy = false
                    }
                }
                .padding(vertical = 12.dp), contentAlignment = Alignment.Center,
        ) { Text(if (busy) "연결 중…" else "지갑으로 연결", color = Color.White, fontWeight = FontWeight.SemiBold, fontSize = 13.sp) }

        if (status.isNotEmpty()) {
            Spacer(Modifier.height(8.dp))
            Text(status, color = if (status.startsWith("연결됨")) GREEN else b.textSecondary,
                fontSize = 11.sp, fontFamily = FontFamily.Monospace)
        }
        if (hubs.value.isNotEmpty()) {
            Spacer(Modifier.height(8.dp))
            hubs.value.forEach { Text("• $it", color = b.textSecondary, fontSize = 11.sp, fontFamily = FontFamily.Monospace) }
        }
    }
}

@Composable
private fun ChoiceChip(label: String, selected: Boolean, onClick: () -> Unit) {
    val b = LocalBrand.current
    Box(
        Modifier.clip(RoundedCornerShape(10.dp))
            .background(if (selected) b.pink else b.pink.copy(alpha = 0.12f))
            .clickableNoRipple { onClick() }.padding(horizontal = 12.dp, vertical = 8.dp),
    ) {
        Text(label, color = if (selected) Color.White else b.pink, fontWeight = FontWeight.SemiBold, fontSize = 12.sp)
    }
}

@Composable
private fun ModeRow(title: String, desc: String, selected: Boolean, onClick: () -> Unit) {
    val b = LocalBrand.current
    Row(
        Modifier.fillMaxWidth().clip(RoundedCornerShape(12.dp))
            .background(if (selected) b.pink.copy(alpha = 0.10f) else Color.Transparent)
            .clickableNoRipple { onClick() }.padding(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.size(20.dp).clip(RoundedCornerShape(50))
            .background(if (selected) b.pink else b.stroke), contentAlignment = Alignment.Center) {
            if (selected) Text("✓", color = Color.White, fontSize = 12.sp)
        }
        Spacer(Modifier.width(12.dp))
        Column(Modifier.weight(1f)) {
            Text(title, color = b.textPrimary, fontSize = 14.sp, fontWeight = FontWeight.SemiBold)
            Text(desc, color = b.textSecondary, fontSize = 11.sp)
        }
    }
}

@Composable
private fun ToggleRow(label: String, checked: Boolean, onChange: (Boolean) -> Unit) {
    val b = LocalBrand.current
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Text(label, color = b.textPrimary, fontSize = 14.sp, modifier = Modifier.weight(1f))
        Switch(checked = checked, onCheckedChange = onChange,
            colors = SwitchDefaults.colors(checkedTrackColor = b.pink))
    }
}

@Composable
private fun MetricBar(label: String, value: Float, valueText: String, color: Color) {
    val b = LocalBrand.current
    Column(Modifier.padding(vertical = 5.dp)) {
        Row {
            Text(label, color = b.textSecondary, fontSize = 12.sp)
            Spacer(Modifier.weight(1f))
            Text(valueText, color = b.textPrimary, fontSize = 12.sp, fontWeight = FontWeight.SemiBold)
        }
        Spacer(Modifier.height(4.dp))
        Box(Modifier.fillMaxWidth().height(8.dp).clip(RoundedCornerShape(4.dp)).background(b.stroke)) {
            Box(Modifier.fillMaxWidth(value.coerceIn(0f, 1f)).fillMaxHeight()
                .clip(RoundedCornerShape(4.dp)).background(color))
        }
    }
}

@Composable
private fun Gauge(label: String, value: Float, valueText: String, color: Color, live: Boolean) {
    val b = LocalBrand.current
    val s = LocalStrings.current
    Column(Modifier.padding(vertical = 6.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(label, color = b.textPrimary, fontSize = 13.sp, fontWeight = FontWeight.Medium)
            if (!live) {
                Spacer(Modifier.width(6.dp))
                Text(s.t("ns.estimated"), color = b.textSecondary, fontSize = 9.sp,
                    modifier = Modifier.background(b.stroke, RoundedCornerShape(4.dp)).padding(horizontal = 4.dp, vertical = 1.dp))
            }
            Spacer(Modifier.weight(1f))
            Text(valueText, color = color, fontSize = 13.sp, fontWeight = FontWeight.Bold, fontFamily = FontFamily.Monospace)
        }
        Spacer(Modifier.height(5.dp))
        Box(Modifier.fillMaxWidth().height(10.dp).clip(RoundedCornerShape(5.dp)).background(b.stroke)) {
            Box(Modifier.fillMaxWidth(value.coerceIn(0f, 1f)).fillMaxHeight()
                .clip(RoundedCornerShape(5.dp)).background(color))
        }
    }
}

File: wallet/ios/App/NodeSettings.swift
import SwiftUI
import Core

/// Mobile-phone node settings: pick a compute backend (MLX GPU / CPU) and mode,
/// see the estimated resource impact, and run the device live with real-time gauges.
struct NodeSettingsView: View {
    @ObservedObject var staking: StakingStore
    @ObservedObject private var loc = Localizer.shared
    @ObservedObject private var agent = AgentControlServer.shared
    @ObservedObject private var participation = HubParticipation.shared
    @State private var stats = DeviceStats()
    @State private var hubURL = "https://hub.kvasir-ai.net"
    @State private var hubStatus = ""
    @State private var hubBusy = false
    @State private var keyStatus = ""
    @State private var keyBusy = false

    private var profile: NodeProfile { nodeProfile(backend: staking.nodeBackend, mode: staking.nodeMode) }

    var body: some View {
        ZStack {
            BrandBackground()
            ScrollView {
                VStack(spacing: 14) {
                    backendCard
                    modeCard
                    infographicCard
                    policyCard
                    hubCard
                    apiKeyCard
                    if staking.nodeLive { liveCard }
                }
                .padding(20)
            }
        }
        .navigationTitle(loc.t("home.nodeSettingsTitle"))
        .navigationBarTitleDisplayMode(.inline)
        .task(id: staking.nodeLive) { await runLiveLoop() }
    }

    // MARK: backend

    private var backendCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(loc.t("nodeSettings.backendTitle"))
                .font(.system(.headline, design: .rounded)).foregroundStyle(Brand.textPrimary)
            Text(loc.t("nodeSettings.backendDesc"))
                .font(.caption).foregroundStyle(Brand.textSecondary)
            HStack(spacing: 8) {
                ForEach(NodeBackend.all, id: \.id) { b in
                    chip(b.label, selected: staking.nodeBackend == b.id) { staking.nodeBackend = b.id }
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    // MARK: connect to a remote hub (SIWS node token -> outbound shard serving)

    // The credit API key is stored on this device and the gateway keeps only its
    // hash, so a key it no longer recognises has to be reminted rather than
    // repaired. This is the way out of "invalid API key" from inside the app.
    private var apiKeyCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(loc.t("apiKey.title"))
                .font(.system(.headline, design: .rounded)).foregroundStyle(Brand.textPrimary)
            Text(loc.t("apiKey.desc"))
                .font(.caption).foregroundStyle(Brand.textSecondary)
            Button {
                guard !keyBusy else { return }
                keyBusy = true
                keyStatus = loc.t("apiKey.working")
                Task {
                    let r = await staking.reissueCreditApiKey()
                    keyStatus = r; keyBusy = false
                }
            } label: {
                Text(keyBusy ? loc.t("apiKey.working") : loc.t("apiKey.reissue"))
                    .fontWeight(.semibold).foregroundStyle(.white)
                    .frame(maxWidth: .infinity).padding(.vertical, 12)
                    .background(Brand.pink.opacity(keyBusy ? 0.4 : 1),
                                in: RoundedRectangle(cornerRadius: 10, style: .continuous))
            }
            .disabled(keyBusy)
            if !keyStatus.isEmpty {
                Text(keyStatus).font(.caption2)
                    .foregroundStyle(keyStatus == loc.t("apiKey.ok") ? Color.green : Brand.textSecondary)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    private var hubCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(loc.t("nodeSettings.hubConnectTitle"))
                .font(.system(.headline, design: .rounded)).foregroundStyle(Brand.textPrimary)
            Text(loc.t("nodeSettings.hubConnectDesc"))
                .font(.caption).foregroundStyle(Brand.textSecondary)
            TextField(loc.t("nodeSettings.hubUrl"), text: $hubURL)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled(true)
                .keyboardType(.URL)
                .font(.callout)
                .padding(12)
                .background(Brand.stroke, in: RoundedRectangle(cornerRadius: 10, style: .continuous))
                .foregroundStyle(Brand.textPrimary)
            Button {
                guard !hubBusy else { return }
                hubBusy = true
                hubStatus = loc.t("nodeSettings.hubSigning")
                Task {
                    let r = await staking.connectHub(url: hubURL)
                    hubStatus = r; hubBusy = false
                }
            } label: {
                Text(hubBusy ? loc.t("nodeSettings.hubConnecting") : loc.t("nodeSettings.hubConnectBtn"))
                    .fontWeight(.semibold).foregroundStyle(.white)
                    .frame(maxWidth: .infinity).padding(.vertical, 12)
                    .background(Brand.pink.opacity(hubBusy ? 0.4 : 1),
                                in: RoundedRectangle(cornerRadius: 10, style: .continuous))
            }
            .disabled(hubBusy)
            if !hubStatus.isEmpty {
                Text(hubStatus).font(.caption2.monospaced())
                    .foregroundStyle(hubStatus.hasPrefix(loc.t("nodeSettings.hubConnected")) ? Color.green : Brand.textSecondary)
            }
            ForEach(participation.hubs, id: \.self) { h in
                Text("• \(h)").font(.caption2).foregroundStyle(Brand.textSecondary)
            }
            if !participation.lastStatus.isEmpty {
                Text(participation.lastStatus).font(.caption2).foregroundStyle(Brand.textSecondary)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    // MARK: mode

    private var modeCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(loc.t("nodeSettings.modeTitle"))
                .font(.system(.headline, design: .rounded)).foregroundStyle(Brand.textPrimary)
            modeRow(loc.t("nodeSettings.modeLocalTitle"), loc.t("nodeSettings.modeLocalDesc"), id: "local_shard")
            modeRow(loc.t("mode.rpcWorker"), loc.t("nodeSettings.modeRpcDesc"), id: "rpc_worker")
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    private func modeRow(_ title: String, _ desc: String, id: String) -> some View {
        let selected = staking.nodeMode == id
        return Button { staking.nodeMode = id } label: {
            HStack(spacing: 12) {
                ZStack {
                    Circle().fill(selected ? Brand.pink : Brand.stroke).frame(width: 22, height: 22)
                    if selected { Image(systemName: "checkmark").font(.caption2.weight(.bold)).foregroundStyle(.white) }
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title).font(.system(.subheadline, design: .rounded).weight(.semibold)).foregroundStyle(Brand.textPrimary)
                    Text(desc).font(.caption2).foregroundStyle(Brand.textSecondary)
                }
                Spacer()
            }
            .padding(12)
            .background(selected ? Brand.pink.opacity(0.10) : .clear, in: RoundedRectangle(cornerRadius: 12, style: .continuous))
        }
        .buttonStyle(.plain)
    }

    // MARK: infographic

    private var infographicCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(alignment: .top) {
                Text(loc.t("nodeSettings.resourceTitle"))
                    .font(.system(.headline, design: .rounded)).foregroundStyle(Brand.textPrimary)
                Spacer()
                VStack(alignment: .trailing, spacing: 0) {
                    Text("\(profile.tokPerSec)")
                        .font(.system(size: 24, weight: .heavy, design: .rounded))
                        .foregroundStyle(Brand.gradient)
                    Text(loc.t("nodeSettings.tokPerSecCaption")).font(.system(size: 10)).foregroundStyle(Brand.textSecondary)
                }
            }
            Text(profile.computeUnit).font(.system(.footnote, design: .rounded).weight(.semibold)).foregroundStyle(Brand.blue)
            metricBar(loc.t("nodeSettings.memImpact"), profile.memImpact, "\(Int(profile.memImpact * 100))%", Brand.blue)
            metricBar(loc.t("nodeSettings.thermal"), profile.thermal, thermalWord(profile.thermal), .orange)
            metricBar(loc.t("nodeSettings.performance"), profile.performance, "\(Int(profile.performance * 100))%", .green)
            Text("※ \(profile.note)").font(.caption2).foregroundStyle(Brand.textSecondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    private func metricBar(_ label: String, _ value: Float, _ valueText: String, _ color: Color) -> some View {
        VStack(spacing: 4) {
            HStack {
                Text(label).font(.caption).foregroundStyle(Brand.textSecondary)
                Spacer()
                Text(valueText).font(.caption.weight(.semibold)).foregroundStyle(Brand.textPrimary)
            }
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    Capsule().fill(Brand.stroke).frame(height: 8)
                    Capsule().fill(color).frame(width: geo.size.width * CGFloat(min(max(value, 0), 1)), height: 8)
                }
            }
            .frame(height: 8)
        }
        .padding(.vertical, 3)
    }

    // MARK: policy + live

    private var policyCard: some View {
        VStack(spacing: 10) {
            Toggle(loc.t("nodeSettings.chargingOnly"), isOn: $staking.nodeChargingOnly)
                .tint(Brand.pink).foregroundStyle(Brand.textPrimary)
                .font(.system(.subheadline, design: .rounded))
            Toggle(loc.t("nodeSettings.runLive"), isOn: $staking.nodeLive)
                .tint(Brand.pink).foregroundStyle(Brand.textPrimary)
                .font(.system(.subheadline, design: .rounded))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    private var liveCard: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(spacing: 8) {
                Circle().fill(.green).frame(width: 9, height: 9)
                Text(loc.t("nodeSettings.liveGauges")).font(.system(.headline, design: .rounded)).foregroundStyle(Brand.textPrimary)
                Spacer()
                Text(stats.charging ? loc.t("nodeSettings.charging") : loc.t("nodeSettings.battery")).font(.caption).foregroundStyle(Brand.textSecondary)
            }
            if staking.nodeMode == "rpc_worker" {
                VStack(alignment: .leading, spacing: 3) {
                    if let paused = staking.nodePausedReason {
                        Label(paused, systemImage: "pause.circle.fill")
                            .font(.caption).foregroundStyle(.orange)
                    } else {
                        Text("RPC worker \(agent.running ? "listening" : "starting") · agent :\(agent.agentPort) · rpc :\(agent.rpcPort)")
                            .font(.system(.caption, design: .monospaced)).foregroundStyle(agent.running ? .green : Brand.textSecondary)
                    }
                    if let c = agent.boundController {
                        Text("bound to \(c)").font(.caption2).foregroundStyle(Brand.textSecondary)
                    }
                    if !agent.lastEvent.isEmpty && staking.nodePausedReason == nil {
                        Text(agent.lastEvent).font(.caption2).foregroundStyle(Brand.textSecondary).lineLimit(2)
                    }
                }
                Label(loc.t("nodeSettings.screenOnHint"), systemImage: "bolt.fill")
                    .font(.caption2).foregroundStyle(Brand.textSecondary)
            }
            gauge("RAM", ramFrac, String(format: "%.1f / %.1f GB", Double(stats.ramUsedMb) / 1024, Double(stats.ramTotalMb) / 1024), Brand.blue, live: true)
            gauge(loc.t("nodeSettings.cpuLoad"), stats.cpuLoad, "\(Int(stats.cpuLoad * 100))%", .green, live: true)
            gauge(loc.t("nodeSettings.thermalState"), stats.thermal, loc.t(stats.thermalWord), .orange, live: true)
            gauge("GPU (\(staking.nodeBackend.uppercased()))", profile.performance, loc.t("nodeSettings.estimated"), Brand.pink, live: false)
            Text(loc.t("nodeSettings.gaugeNote")).font(.system(size: 10)).foregroundStyle(Brand.textSecondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .brandCard()
    }

    private var ramFrac: Float {
        stats.ramTotalMb > 0 ? Float(stats.ramUsedMb) / Float(stats.ramTotalMb) : 0
    }

    private func gauge(_ label: String, _ value: Float, _ valueText: String, _ color: Color, live: Bool) -> some View {
        VStack(spacing: 5) {
            HStack(spacing: 6) {
                Text(label).font(.system(.subheadline, design: .rounded).weight(.medium)).foregroundStyle(Brand.textPrimary)
                if !live {
                    Text(loc.t("nodeSettings.estimated")).font(.system(size: 9)).foregroundStyle(Brand.textSecondary)
                        .padding(.horizontal, 4).padding(.vertical, 1)
                        .background(Brand.stroke, in: RoundedRectangle(cornerRadius: 4))
                }
                Spacer()
                Text(valueText).font(.system(.subheadline, design: .monospaced).weight(.bold)).foregroundStyle(color)
            }
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    Capsule().fill(Brand.stroke).frame(height: 10)
                    Capsule().fill(color).frame(width: geo.size.width * CGFloat(min(max(value, 0), 1)), height: 10)
                }
            }
            .frame(height: 10)
        }
        .padding(.vertical, 6)
    }

    // MARK: helpers

    private func chip(_ label: String, selected: Bool, _ action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(label)
                .font(.caption.weight(.semibold))
                .foregroundStyle(selected ? .white : Brand.pink)
                .padding(.horizontal, 12).padding(.vertical, 8)
                .background(selected ? Brand.pink : Brand.pink.opacity(0.12), in: RoundedRectangle(cornerRadius: 10, style: .continuous))
        }
        .buttonStyle(.plain)
    }

    private func thermalWord(_ v: Float) -> String {
        loc.t(v >= 0.75 ? "thermal.high" : (v >= 0.5 ? "thermal.medium" : "thermal.low"))
    }

    /// While live, register with the current profile then heartbeat + refresh gauges.
    /// The rpc_worker data plane is NOT tied to this view — StakingStore owns the
    /// worker's lifecycle so it survives leaving the screen.
    private func runLiveLoop() async {
        guard staking.nodeLive else { return }
        // Registration, heartbeat and charge-driven pause/resume are owned app-wide
        // by StakingStore (so the node stays online off this screen and across
        // background/foreground); this view loop only refreshes the on-screen gauges.
        while staking.nodeLive && !Task.isCancelled {
            stats = NodeTelemetry.shared.read()
            try? await Task.sleep(for: .seconds(1.5))
        }
    }
}
